src/herder/Upgrades.cpp

```diff
@@ -318,4 +328,237 @@ Upgrades::timeForUpgrade(uint64_t time) const
+    db.getEntryCache().clear();
+    auto offersByAccount = OfferFrame::loadAllOffers(db);
+
+    for (auto& accountOffers : offersByAccount)
+    {
+        std::map<Asset, std::unique_ptr<int64_t>> initialBuyingLiabilities;
+        std::map<Asset, std::unique_ptr<int64_t>> initialSellingLiabilities;
+        for (auto const& offerFrame : accountOffers.second)
+        {
+            auto const& offer = offerFrame->getOffer();
+            addLiabilities(initialBuyingLiabilities, offer.sellerID,
+                           offer.buying, offerFrame->getBuyingLiabilities());
+            addLiabilities(initialSellingLiabilities, offer.sellerID,
+                           offer.selling, offerFrame->getSellingLiabilities());
+        }
+
+        auto accountFrame =
+            AccountFrame::loadAccount(ld, accountOffers.first, db);
+        auto& account = accountFrame->getAccount();
+        assert(accountFrame);
+        int64_t balanceAboveReserve =
+            accountFrame->getBalanceAboveReserve(ledgerManager);
+
+        std::map<Asset, Liabilities> liabilities;
+        for (auto const& offerFrame : accountOffers.second)
+        {
+            auto const& offer = offerFrame->getOffer();
+            bool erase = shouldDeleteOffer(
+                offer.sellerID, offer.selling, balanceAboveReserve,
+                initialSellingLiabilities, getAvailableBalance, ld, db);
+            erase = erase || shouldDeleteOffer(offer.sellerID, offer.buying,
+                                               balanceAboveReserve,
+                                               initialBuyingLiabilities,
+                                               getAvailableLimit, ld, db);
+
+            if (erase)
+            {
+                accountFrame->addNumEntries(-1, ledgerManager);
+                offerFrame->storeDelete(ld, db);
+            }
+            else
+            {
+                if (offer.buying.type() == ASSET_TYPE_NATIVE ||
+                    !(offer.sellerID == getIssuer(offer.buying)))
+                {
+                    bool r =
+                        stellar::addBalance(liabilities[offer.buying].buying,
+                                            offerFrame->getBuyingLiabilities());
+                    assert(r);
+                }
+
+                if (offer.selling.type() == ASSET_TYPE_NATIVE ||
+                    !(offer.sellerID == getIssuer(offer.selling)))
+                {
+                    bool r = stellar::addBalance(
+                        liabilities[offer.selling].selling,
+                        offerFrame->getSellingLiabilities());
+                    assert(r);
+                }
+            }
+        }
+
+        for (auto const& assetLiabilities : liabilities)
+        {
+            Asset const& asset = assetLiabilities.first;
+            Liabilities const& liab = assetLiabilities.second;
+            if (asset.type() == ASSET_TYPE_NATIVE)
+            {
+                account.ext.v(1);
```

Comment:
call `addBuyingLiabilities` and `addSellingLiabilities` to avoid messing with the `ext.v1()` syntax outside of `AccountFrame`

Reply:
same for trustline below